docs/data/data-grid/server-side-data/row-grouping.md

```diff
@@ -2,14 +2,90 @@
+
+```tsx
+const customDataSource: GridDataSource = {
+  getRows: async (params) => {
+    // Fetch the data from the server
+  },
+  getGroupKey: (row) => {
+    // Return the group key for the row, e.g. `name`
+    return row.name;
+  },
+  getChildrenCount: (row) => {
+    // Return the number of children for the row
+    return row.childrenCount;
+  },
+};
+```
 
-👍 Upvote [issue #10859](https://github.com/mui/mui-x/issues/10859) if you want to see it land faster.
+In addition to `groupKeys`, the `getRows()` callback receives a `groupFields` parameter. This corresponds to the current `rowGroupingModel`. Use `groupFields` on the server to group the data for each `getRows()` call.
 
-Don't hesitate to leave a comment on the same issue to influence what gets built. Especially if you already have a use case for this component, or if you are facing a pain point with your current solution.
+```tsx
+const getRows: async (params) => {
+  const urlParams = new URLSearchParams({
+    // Example: JSON.stringify(['20th Century Fox', 'James Cameron'])
+    groupKeys: JSON.stringify(params.groupKeys),
+    // Example: JSON.stringify(['company', 'director'])
+    groupFields: JSON.stringify(params.groupFields),
+  });
+  const getRowsResponse = await fetchRows(
+    // Server should group the data based on `groupFields` and
+    // extract the rows for the nested level based on `groupKeys`
+    `https://mui.com/x/api/data-grid?${urlParams.toString()}`,
+  );
+  return {
+    rows: getRowsResponse.rows,
+    rowCount: getRowsResponse.rowCount,
+  };
+}
+```
+
+{{"demo": "ServerSideRowGroupingDataGrid.js", "bg": "inline"}}
+
+:::warning
+For complex data, consider using `colDef.groupingValueGetter` to extract the grouping value. This value is passed in the `groupKeys` parameter when `getRows` is called.
+
+Ensure your backend can interpret the `groupKeys` parameter generated by `colDef.groupingValueGetter` to retrieve grouping values for child rows.
 :::
+
+## Error handling
+
+If an error occurs during a `getRows` call, the Data Grid displays an error message in the row group cell. `unstable_onDataSourceError` is also triggered with the error and the fetch params.
+
+This example shows error handling with toast notifications and default error messages in grouping cells. Caching is disabled for simplicity.
+
+{{"demo": "ServerSideRowGroupingErrorHandling.js", "bg": "inline"}}
+
+## Group expansion
+
+The group expansion works similar to the [data source tree data](/x/react-data-grid/server-side-data/tree-data/#group-expansion).
+The following demo uses `defaultGroupingExpansionDepth='-1'` to expand all the groups.
+
+{{"demo": "ServerSideRowGroupingGroupExpansion.js", "bg": "inline"}}
+
+## Demo
+
+In the following demo, use the auto generated data based on the `Commodities` dataset to simulate the server-side row grouping.
+
+{{"demo": "ServerSideRowGroupingFullDataGrid.js", "bg": "inline"}}
```

Comment:
Any ideas about what is causing these lags? Maybe client-side filtering kicks in?


https://github.com/user-attachments/assets/48f6190a-096e-479b-83eb-dd7ffc3e93fb

Reply:
Thank you @cherniavskii for pointing it out. It really does look bad.

Filtering logic on the `useMockServer` side isn't very optimal atm. Results on 20k rows:

![image](https://github.com/user-attachments/assets/16d926ac-b2d3-48f6-a64f-14777ec187fa)

We'll have to replicate/reuse the frontend filtering engine there somehow to be as performant. (I guess the same mock server computation bottle-neck was a rationale for [limiting](https://github.com/mui/mui-x/blob/61bef1e4cfdc8c026eb73b7ce2887a1d459d444b/packages/x-data-grid-generator/src/services/tree-data-generator.ts#L44) tree data max row size to `1000`)

For now, I [limited](https://github.com/mui/mui-x/pull/13826/commits/d849c385fc405d485175c9104eb10ebe6638670e) the data size for this demo, it's smoother. If needed, we could try to optimize the backend logic sometime in the future.

Reply:
Oh, we didn't anticipate that the mock server wouldn't be performant enough for large datasets.
At this point, I feel like it's better to set up a backend with a database for it.

Reply:
> I feel like it's better to set up a backend with a database for it.

Is it worth the effort though? Technically we could replicate the same filtering engine on the backend which we have on the client app.

Even demonstrating on a lower data size should be fine for now, since the users should already have their backend servers setup.

Is there a specific aspect in your mind we could not cover by a dummy server?

Reply:
> Is it worth the effort though? Technically we could replicate the same filtering engine on the backend which we have on the client app.

I think the effort would be smaller than making `useMockServer` work and perform well. Plus, users could use this example as a reference when integrating with their DB, as opposed to `useMockServer` which is an internal implementation that is not useful to anyone integrating Data Grid with their backend.

> Is there a specific aspect in your mind we could not cover by a dummy server?

My main concern is that we're trying to build data transformation from scratch, while we can use a database query that is optimized for it already – e.g. for row grouping, we can use groupBy + where instead of a ton of custom code: https://github.com/mui/mui-x/pull/13826/files#diff-97d37099273d5d313dc00f50320ced3b902a089d924b2092213d2573ee5c63c8

Reply:
Your points make sense 👍

I am more concerned about the BE server infra management (server, db, etc) in this case that'd support high bandwidth without causing disruptions. I'm not sure if it will be possible using toolpad, even if it is, we have used it for internal tools for now, not on the public documentation, or have we? 🤔

In any case, let's take it as a separate issue: https://github.com/mui/mui-x/issues/15098

Reply:
> I am more concerned about the BE server infra management (server, db, etc) in this case that'd support high bandwidth without causing disruptions.

Yes, this might be challenging indeed.